Fix #8087-Unable to add new web service client

pepness opened the pull request:
[Fix issue 8087](https://github.com/apache/netbeans/issues/8087)
NetBeans Notes:
- Add support for Jakarta EE 9/10 plugins for metro, jax-ws, and maven-war-plugin
- Add support for jakarta namespaces for Jakarta EE 9 and later
- Change jax-ws group id to `com.sun.xml.ws`
- Add instance variables to `JaxWsClientCreator` that will help with the new logic to support jakarta namespaces and reduce method calls
- Refactor code and add missing add support for creating web service clients for EJB, CAR and Java SE maven
projects
- Simplify logic that compute whether a project is Web, Car, Jar, EJB
- Add missing EJB version support
- Add missing Servlet version support
- Add missing AppClient descriptor for Jakarta EE 11
- Add missing AppClient resolvers Java EE 7 to Jakarta EE 11
- Add new method getProfile()
- Change logic in getJ2eeProfile() and use the new method getProfile()
- Remove version model when calling addJaxWSPlugin(), the default version will vary for the computed Profile, Java EE 8 will use model 2.2,  Jakarta EE 9/10/11 will use 3.0
- Generate new code for equals() and hashCode() methods
- Use Pattern Matching for instanceof
- Use Switch Expressions
- Use constants when possible
- Use diamond inference
- Use try-with-resources
- Increase array size

NetBeans Testing:

- Verify successful execution of libraries and licenses Ant test
- Verify successful execution of Verify Sigtests
- Verify successful execution of unit tests for modules `maven.jaxws`
- Started NetBeans and ensure the log didn't have any ERROR or new WARNINGS
- Successfully register GlassFish 5/6/7
  - Create a Jakarta EE 8/9/10 maven web app and create a new `Web Service from WSDL` and `Web Service Client`
  - Create a Jakarta EE 8/9/10 maven ejb app and create a new `Web Service from WSDL` and `Web Service Client`
  - Create a maven Java App with Java 8/11/25 and create a new `Web Service from WSDL` and `Web Service Client`
    - Verify that it creates a new Web Service
    - Verify that the generated sources are using the correct version plugins
    - Verify that the generated sources are using the correct namespace

Jakarta EE 8
![Screenshot from 2025-02-10 15-19-47](https://github.com/user-attachments/assets/30f56290-6bfe-4539-9740-8debb52e6f7d)
Jakarta EE 9
![Screenshot from 2025-02-10 15-19-40](https://github.com/user-attachments/assets/20d4f363-f98e-4f90-9559-05f929ef59a7)
Jakarta EE 10
<img width="2170" height="906" alt="Screenshot From 2026-05-11 18-04-36" src="https://github.com/user-attachments/assets/cd75f291-018b-449f-9efb-baa211df8208" />
Tested with multiple projects (Web, EJB, JavaSE) with different Java/JakartaEE versions
<img width="2206" height="1009" alt="Screenshot From 2026-05-11 19-13-11" src="https://github.com/user-attachments/assets/5bf5489d-ac69-402d-b86f-bdc3b2931052" />

matthiasblaesing commented on enterprise/maven.jaxws/src/org/netbeans/modules/maven/jaxws/MavenModelUtils.java:
```diff
@@ -59,14 +58,28 @@
+    
+    public static final String MAVEN_PLUGINS_GROUP_ID = "org.apache.maven.plugins"; //NOI18N
+    public static final String WAR_PLUGIN_ARTIFACT_ID = "maven-war-plugin"; //NOI18N
+    public static final String WAR_PLUGIN_VERSION = "2.3.4"; //NOI18N
```
I would bump to `3.4.0` if there is not a very good reason to stay on the old version. I remember that I bumped several projects to a 3.X version as 2.X failed when used with JDK 21+

pepness replied:
Bump to latest version 3.5.1

matthiasblaesing commented on enterprise/maven.jaxws/src/org/netbeans/modules/maven/jaxws/MavenModelUtils.java:
```diff
@@ -59,14 +58,28 @@
  */
 public final class MavenModelUtils {
 
+    private static Profile profile;
```
This will become a problem. If you have one maven project with JakartaEE and one with JavaEE open, both will use the same `profile`. You will need to pass the "isJakarta" property (or the profile) from the external entry points, to decide per call.

A passing question: is it save to assume this is only called for JavaEE/JakarataEE projects? If it is called for JavaSE projects, there will be no profile.

pepness replied:
Java EE 8 will use model 2.2, Jakarta EE 9/10/11 will use 3.0, tested with Maven Web, EJB, and JavaSE projects with different Java and Java/Jakarta EE versions, each call will compute the correct API to use.

matthiasblaesing reviewed (changes requested):
Thank you! In general looks sane to me. I left inline comments.

pepness committed:
Fix #8087 - Unable to add new web service client

-Add missing EJB version support
-Add missing Servlet version support
-Add missing AppClient descriptor for Jakarta EE 11
-Add missing AppClient resolvers Java EE 7 to Jakarta EE 11
-Replace usage of deprecated methods from codehaus
-Add missing support for Client Applications
-Add a new module dependency in order to get the source level of a java
 file
-Add module 'maven.jaxws' as a friend
-Implement missing interface and its methods
-Move method 'getProfileFromPOM' to 'MavenProjectSupport' class and use
 it when computing the profile in all implementations.
-Add new method 'getSourceLevel'
-Add 'getSourceLevel' method to 'MavenProjectSupport', useful with JSE
 maven projects
-Add new method 'getProfile' that compute the profile in the classes
 that create the WS Client; 'JaxWsClientCreator' and
 'JaxWsServiceCreator'
-Change logic in 'getJ2eeProfile' and use the new method 'getProfile'
-Generate new code for 'equals' and 'hashCode' methods
-Remove version model when calling 'addJaxWSPlugin', the default version
 will vary for the computed 'Profile', Java EE 8 will use model 2.2,
 Jakarta EE 9/10/11 will use 3.0.
-Add support for creating web services for EJB, CAR and Java SE maven
 projects.
-Simplify logic that compute whether a project is Web, Car, Jar, EJB
-Add support for Car and Jar projects to 'isJakartaEENameSpace'
-Fix wrong artifact and update GlassFish 7/8 version
-Update maven coordinates inside test
-Fix mismatch when using 'MavenProjectSupport.getProfileFromPOM' to
 retrieve the correct 'Profile'.
-'equals' and 'hashCode' methods should cover the same fields
-Remove the redundant Xpp3Dom variable
-Bump 'jaxws-maven-plugin' version to 4.0.4
-Bump 'webservices-api' & 'webservices-rt' version to 4.0.6
-Bump 'maven-war-plugin' version to 3.5.1

Cleaning/Refactoring:
-Bump javac to version 17 and use:
 -Pattern Matching for instanceof
 -Switch Expressions
-Use try-with-resources
-Add @Override annotation
-Remove unused imports
-Use diamond operator
-Use parseBoolean over valueOf
-Use Multi-catch
-Add Logger and use it
-Replace usage of deprecated methods from codehaus
-Add useful comment for debugging
-Add useful variable for debugging
-Use constants where makes sense

mbien commented on enterprise/maven.jaxws/src/org/netbeans/modules/maven/jaxws/WSUtils.java:
```diff
@@ -137,24 +141,10 @@ public static void generateSunJaxwsFile(final FileObject targetDir) throws IOExc
-                try {
-                    os = sunJaxwsFo.getOutputStream(lock);
-                    osw = new OutputStreamWriter(os, StandardCharsets.UTF_8);
-                    bw = new BufferedWriter(osw);
+                try (FileLock lock = sunJaxwsFo.lock(); OutputStream os = sunJaxwsFo.getOutputStream(lock);
+                        OutputStreamWriter osw = new OutputStreamWriter(os, StandardCharsets.UTF_8);
+                        BufferedWriter bw = new BufferedWriter(osw)) {
                     bw.write(sunJaxwsContent);
-                } finally {
```
suggestion:
```java
                try (FileLock lock = sunJaxwsFo.lock(); 
                     BufferedWriter bw = new BufferedWriter(
                           new OutputStreamWriter(sunJaxwsFo.getOutputStream(lock), StandardCharsets.UTF_8))) {
```
`bw` will close all wrapped streams.

mbien reviewed (commented):
left some comments while looking through it - nice cleanups!

(please don't forget to squash btw)

pepness commented:
> @pepness what is the state of this? I think it would be good to rebase this onto current master, cleanup the history and check if all comments are addressed. The static `profile` variable is still there and globals like these are really ugly to debug.

I updated some maven plugins and removed the use of the static variable. Will re-test everything and finish by tomorrow.

pepness commented:
All done.

matthiasblaesing commented on enterprise/maven.jaxws/src/org/netbeans/modules/maven/jaxws/MavenModelUtils.java:
```diff
@@ -689,14 +688,12 @@ private static String getJaxWsVersion(Profile profile) {
-     * @see hasJaxWsAPI
      */
-    private static void getProfile(Project project) {
+    public Profile getProfile(Project project) {
```
Unless I'm mistaken, `getProfile` can stay static as there is no state other than what is stored in `project` that is passed in. The instances created in the `*Creator` classes can be dropped and replaced with a static call.

pepness replied:
Good call, done.

matthiasblaesing reviewed (approved):
Looks sane to me.

Final wish: Please check if the number of commits could be reduced. Should there be a need to revert this after merge 24 commits are really ugly to remove.

Apart from that I think merging this now, i.e. early in the 31 cycle is a very good idea.

Thank you!

pepness commented:
All done.

> Looks sane to me.
> 
> Final wish: Please check if the number of commits could be reduced. Should there be a need to revert this after merge 24 commits are really ugly to remove.
> 
> Apart from that I think merging this now, i.e. early in the 31 cycle is a very good idea.
> 
> Thank you!

Rebase and squash it to 10 commits.

matthiasblaesing commented:
Unless I'm mistaken nothing in here touches the sig files (the most probable cause for merge conflicts after a release), so my perspective this can be merged.

pepness commented:
> Unless I'm mistaken nothing in here touches the sig files (the most probable cause for merge conflicts after a release), so my perspective this can be merged.

Should I squash everything to 1 commit?

matthiasblaesing commented:
> > Unless I'm mistaken nothing in here touches the sig files (the most probable cause for merge conflicts after a release), so my perspective this can be merged.
> 
> Should I squash everything to 1 commit?

Good question. I depends on whether the commit individually make sense or should be separated. For example if code cleanup and feature work are both present in a PR I prefer separate commits, so that later it is clear what was the intention. On the other hand a single commit is easier to remove should it become necessary.

I thing there are sensible options for squashing. [57b9a65](https://github.com/apache/netbeans/pull/8228/commits/57b9a65a6a42b8e5abe6527fe37e2e7807c4c48f) looks like a candidate to keep separate, but could also be part of a single big change commit.

Does this help you?

mbien commented:
fyi: commit msg is now missing the first line (it is starting with a bullet point), this might not be intended

edit: usually its the pr title or a one sentence summary, followed by blank line

pepness commented:
> fyi: commit msg is now missing the first line (it is starting with a bullet point), this might not be intended
> 
> edit: usually its the pr title or a one sentence summary, followed by blank line

You are right, I fixed the commit message.

matthiasblaesing commented:
Thanks

matthiasblaesing commented:
@pepness would you mind merging this unless you think it is not ready yet?
